### PR TITLE
Introduce `target_columns` for `replacements`

### PR DESCRIPTION
**Pull request by Adel-Moumen**
## What does this PR do?

This PR adds the feature of `target_columns` to make sure that `replacements` is only applied on those columns/keys. By default, our `from_csv` method was applying `replacements` on all matching values. In most cases, you might want to only consider a column such as `wav` for instance. 

This PR will also serve as a way to make sure that most recipes fully benefits from `replacements`. Indeed, all the recipes are using `replacements={"data_root": data_folder}` but nothing is happening because the data_preparation script is not aware of this and never replace the audio path with `$data_root`. Therefore, the output csvs are in the form `/my/audio/path/file.wav` instead of `$data_root/file.wav` which means that if you are in an HPC such as compute canada or jean zay you'll need to re-run data preparation for nothing. This is critical on dataset that can be very large such as GigaSpeech. 

Note: this PR is a replacement of #2125 that was not suited for every recipes in SB. 

<details>
  <summary><b>Before submitting</b></summary>

- [ ] Did you read the [contributor guideline](https://speechbrain.readthedocs.io/en/latest/contributing.html)?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing [tests](https://github.com/speechbrain/speechbrain/tree/develop/tests) pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Does your code adhere to project-specific code style and conventions?

</details>

## PR review

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
- [ ] Confirm that the changes adhere to compatibility requirements (e.g., Python version, platform)
- [ ] Review the self-review checklist to ensure the code is ready for review

</details>

### CODE REVIEW

File: speechbrain/dataio/dataio.py [hunk 1]
```diff
@@ -41,6 +43,9 @@ def load_data_json(json_path, replacements={}):
     replacements : dict
         (Optional dict), e.g., {"data_folder": "/home/speechbrain/data"}.
         This is used to recursively format all string values in the data.
```
Comment (asumagic) [inline speechbrain/dataio/dataio.py:45]: I know this is not new in the PR but while the way replacements are used is hinted at in the example, it's never actually explained in text. IMO it is also unclear what recursion meant (whether it was somehow related to the replacement or the data structure being transformed).

```suggestion
    replacements : dict
        (Optional dict), e.g., {"data_folder": "/home/speechbrain/data"}.
        Replacements will replace text of the form `{key}` with the
        associated value.
        Replacements are applied recursively to all nested values in the
        input. This excludes keys in dictionaries.
```

File: speechbrain/dataio/dataio.py [hunk 2]
```diff
@@ -66,32 +71,34 @@ def load_data_json(json_path, replacements={}):
...
+                    data[key] = item.format_map(replacements)
+                # If not dict, list or str, do nothing
     if isinstance(data, list):
         for i, item in enumerate(data):
-            if isinstance(item, dict) or isinstance(item, list):
-                _recursive_format(item, replacements)
-            elif isinstance(item, str):
-                data[i] = item.format_map(replacements)
-            # If not dict, list or str, do nothing
+            if target_columns is None or i in target_columns:
```
Comment (asumagic) [inline speechbrain/dataio/dataio.py:93]: Does this functionality even make sense to have for list indices? This seems like unintended behavior and ambiguous especially with nested structures.

File: speechbrain/dataio/dataio.py [hunk 3]
```diff
@@ -66,32 +71,34 @@ def load_data_json(json_path, replacements={}):
...
+                if isinstance(item, dict) or isinstance(item, list):
+                    _recursive_format(item, replacements)
+                elif isinstance(item, str):
+                    data[i] = item.format_map(replacements)
+                # If not dict, list or str, do nothing
 
 
-def load_data_csv(csv_path, replacements={}):
+def load_data_csv(csv_path: str, replacements: Dict = {}, target_columns: Union[List, None] = None):
```
Comment (asumagic) [inline speechbrain/dataio/dataio.py:101]:
```suggestion
def load_data_csv(csv_path: str, replacements: Dict[str, str] = {}, target_columns: Union[List, None] = None):
```

File: speechbrain/dataio/dataset.py [hunk 4]
```diff
@@ -403,23 +403,23 @@ def batch_shuffle(self, batch_size):
...
-        data = load_data_csv(csv_path, replacements)
+        data = load_data_csv(csv_path, replacements, target_columns)
         return cls(data, dynamic_items, output_keys)
 
     @classmethod
     def from_arrow_dataset(
-        cls, dataset, replacements={}, dynamic_items=[], output_keys=[]
+        cls, dataset, replacements={}, target_columns = None, dynamic_items=[], output_keys=[]
     ):
         """Loading a prepared huggingface dataset"""
```
Comment (asumagic) [inline speechbrain/dataio/dataset.py:424]: Type annotations + argument docstrings?

File: speechbrain/dataio/dataio.py [hunk 5]
```diff
@@ -31,7 +33,7 @@
 logger = logging.getLogger(__name__)
 
 
-def load_data_json(json_path, replacements={}):
+def load_data_json(json_path: str, replacements: Dict = {}, target_columns: Union[List, None] = None):
```
Comment (asumagic) [inline speechbrain/dataio/dataio.py:36]: From what I could see we are lacking unit tests for the `load_data_csv`/`load_data_json`, I feel like those should be added and should test the column filtering functionality.

**Review by asumagic — CHANGES_REQUESTED** [review]

File: speechbrain/dataio/dataset.py (hunk 4 above)
Comment (Adel-Moumen) [reply speechbrain/dataio/dataset.py:424]: yes will do

**Pull request closed without merging**